### ISSUE TITLE
Pyro models with latent discrete variables / enumeration

### ISSUE THREAD

**Issue by AlexanderAivazidis** [scverse/scvi-tools#1526]
Hi,
Is it possible to train pyro models with latent discrete variables with scvi-tools? Here are examples and explanations of what I mean: https://pyro.ai/examples/enumeration.html
Looking at the PyroSviTrainMixin class it looks to me that this is not implemented. If it doesn't exist, I could probably help with implementing this.

**Comment by adamgayoso** [comment]
Can you give an example of what you'd like to add? The code should be flexible enough now to enable developers to add `TraceEnum_ELBO` etc.

**Comment by AlexanderAivazidis** [comment]
Below is a minimum example of a Gaussian mixture model with 100 cells, 1 gene and 2 components implemented in pyro. I think there are 3 things needed to train models like this in scvi:
1.) add @config_enumerate to the model as in line 22 (hopefully easily possible by adding @config_enumerate to the forward() method of a module)
2.) hide discrete parameters when constructing an autoguide, like in line 39 ("poutine.block(model, hide=['assignment'])")
3.) use "TraceEnum_ELBO" like in line 44

`import numpy as np
import matplotlib.pyplot as plt
import torch
import pyro
import pyro.distributions as dist
from pyro import poutine
from pyro.infer.autoguide import AutoLowRankMultivariateNormal
from pyro.infer.autoguide.initialization import init_to_value
from pyro.infer import SVI, config_enumerate, TraceEnum_ELBO, infer_discrete

# Produce some artificial (normalized) gene expression data:         
n_components = 2
n_genes = 1
n_cells = 100

mus_k = torch.tensor(np.array((5.,20.)), dtype = torch.float32)
weights_k = torch.tensor(np.ones(n_components)/n_components, dtype = torch.float32)
k_c = pyro.sample('k_c', dist.Categorical(weights_k).expand([n_cells]).to_event(1))
data = pyro.sample("obs", dist.Normal(loc = mus_k[k_c], scale = 1).expand([n_cells]).to_event(1))

# Define 1D Gaussian mixture model with two components
@config_enumerate
def model(data):
    
    n_cells = len(data)
    n_components = 2
    scale = 1
    
    # Global variables.
    weights = pyro.sample('weights', dist.Dirichlet(torch.ones(n_components)))
    locs = pyro.sample('locs', dist.Normal(10., 10.).expand([n_components]).to_event(1))

    # Local variables.
    with pyro.plate('cells', n_cells):
        assignment = pyro.sample('assignment', dist.Categorical(weights))
        pyro.sample('obs', dist.Normal(locs[assignment], scale), obs=data)

# Make appropriate auto guide (and hiding discrete variable)
global_guide = AutoLowRankMultivariateNormal(poutine.block(model, hide=['assignment']),
                        init_to_value(values = {'locs': torch.tensor(np.array((1.,10.)), dtype = torch.float32)}))

# Train model:
optim = pyro.optim.Adam({'lr': 0.1, 'betas': [0.8, 0.99]})
elbo = TraceEnum_ELBO(max_plate_nesting=1)
svi = SVI(model, global_guide, optim, loss=elbo)
for i in range(200):
    loss = svi.step(data)

# Predict components:
guide_trace = poutine.trace(global_guide).get_trace(data)  # record the globals
trained_model = poutine.replay(model, trace=guide_trace)  # replay the globals
def classifier(data, temperature=0):
    inferred_model = infer_discrete(trained_model, temperature=temperature,
                                    first_available_dim=-2)  # avoid conflict with data plate
    trace = poutine.trace(inferred_model).get_trace(data)
    return trace.nodes["assignment"]["value"]
k_c_inferred = classifier(data)

# Plot ground truth and inferred components:
scatter = plt.scatter(range(len(data)), data, c = k_c_inferred)
plt.xlabel('cell')
plt.title('Simulated Data')
plt.ylabel('gene expression')
plt.legend(handles=scatter.legend_elements()[0], 
           title="inferred component")
plt.show()
scatter = plt.scatter(range(len(data)), data, c = k_c)
plt.title('Simulated Data')
plt.xlabel('cell')
plt.ylabel('gene expression')
plt.legend(handles=scatter.legend_elements()[0], 
           title="ground truth component")`
![image](https://user-images.githubusercontent.com/46488185/165277612-1fc599a0-4e5e-4f8d-bbd2-970f4c53c387.png)

**Comment by adamgayoso** [comment]
To me, it seems like all the things you requested are currently capable with minimal overhead.

1. config enumerate can be used internally when developers inherit `PyroBaseModuleClass`
2. Same for the blocking
3. Devs can use any Pyro Elbo they'd like when they create the `PyroTrainingPlan`. 


@vitkl what do you think?

**Comment by AlexanderAivazidis** [comment]
Thank you for those pointers. I will implement the minimal Gaussian Mixture model with scvi-tools as a test case. I will let you know how it works out.

**Comment by AlexanderAivazidis** [comment]
Thanks I can confirm it is indeed easily possible to implement something like a Gaussian mixture model in scvi-tools. As a reminder @config_enumerate has to be added to the foward() method of the module. I forgot that initially and it gave non-nonsensical results without giving an error message...

**Comment by vitkl** [comment]
I used the approach @AlexanderAivazidis proposed - which requires all 3 modifications which are all easy to make.
